### PR TITLE
Add unit testing and move to Alpine

### PR DESCRIPTION
**Pull request by fillup**
- Fairly significant code restructuring to use functions so that unit testing could be implemented.
- Unit tests added to cover image name parsing and new task definition creation
- Updated Dockerfile to use Alpine, reducing resulting image size by approximately 200Mb. (From 291Mb to 94Mb).
- Connected with Codeship for automated execution of tests in Docker image

Due to the amount of changes going in here I'd like to request as many people test this version before I merge it in. Tagging all contributors here, but anyone else is invited to test as well. 

@hnryjms @reidab @gregwebs @rafaelfelix @kevinkarwaski @d-me-tree @steevel @johnallen3d @beeva-javiermartincaro @erickrawczyk @brycereitano @mirague @ligaz @tpett @asotirov @gerar2 @inakiabt @donaldpiret @timrc @diddeb @jepettoruti @jangrewe @argusua @pkchesterton

### CODE REVIEW

**Comment by jangrewe** [comment]
I can test tomorrow, but: we've been using ecs-deploy in an Alpine-based image since the beginning, so it's definitely working fine there ;-)

I'll update our image tomorrow when i'm back in the office and will report back.

File: ecs-deploy [hunk 1]
```diff
@@ -1,408 +1,467 @@
...
 
-          exit 0
-        fi
-      fi
+            sleep $every
+            i=$(( $i + $every ))
+        done
+
+        # Timeout
+        echo "ERROR: New task definition not running within $TIMEOUT seconds"
```
Comment (hnryjms) [inline ecs-deploy]: How would you feel about moving the polling code to an additional function? I don't think we'd need to implement it now, but that would give us a quick way to just say `--disable-run-check` or some type of flag.

File: ecs-deploy [hunk 2]
```diff
@@ -1,408 +1,467 @@
...
+                                OUTDATED_REVISION_ARN=$(echo "$TASK_REVISIONS" | jq -r ".taskDefinitionArns[$i]")
 
-            NUM_ACTIVE_REVISIONS=$(echo "$TASK_REVISIONS" | jq ".taskDefinitionArns|length")
+                                echo "Deregistering outdated task revision: $OUTDATED_REVISION_ARN"
 
-            if [[ $NUM_ACTIVE_REVISIONS -gt $MAX_DEFINITIONS ]]; then
-                LAST_OUTDATED_INDEX=$(($NUM_ACTIVE_REVISIONS - $MAX_DEFINITIONS - 1))
-                for i in $(seq 0 $LAST_OUTDATED_INDEX); do
-                    OUTDATED_REVISION_ARN=$(echo "$TASK_REVISIONS" | jq -r ".taskDefinitionArns[$i]")
+                              $AWS_ECS deregister-task-definition --task-definition "$OUTDATED_REVISION_ARN" > /dev/null
```
Comment (hnryjms) [inline ecs-deploy]: As for extraction here with the deregister outdated definitions code I wrote a while ago... we could try

```js
function taskRevisionFamilyPrefix() {
    FAMILY_PREFIX=${TASK_DEFINITION##*:task-definition/}
    FAMILY_PREFIX=${FAMILY_PREFIX%*:[0-9]*}

    if [ "$BASH_SOURCE" != "$0" ]; then
        echo $FAMILY_PREFIX
    fi
}
```

And integration test that??? Maybe???? As for testing the later functionality, we would have to use bash function arguments, which is really meh. I could look into writing the tests for this chunk in the future, after the test framework code is merged.

File: ecs-deploy [hunk 3]
```diff
@@ -1,408 +1,467 @@
...
+# - port
+# - repo
+# - image
+# - tag
+# If a group is missing it will be an empty string
+imageRegex="^([a-zA-Z0-9.\-]+):?([0-9]+)?/([a-zA-Z0-9._-]+)/?([a-zA-Z0-9._-]+)?:?([a-zA-Z0-9\._-]+)?$"
 
 function usage() {
-    set -e
+    #set -e
```
Comment (fillup) [inline ecs-deploy]: good catch

File: ecs-deploy [hunk 4]
```diff
@@ -1,408 +1,467 @@
...
+              AWS_ECS="$AWS_ECS --profile $AWS_PROFILE"
+    fi
 
-# Loop through arguments, two at a time for key and value
-while [[ $# -gt 0 ]]
-do
-    key="$1"
+    if [ $VERBOSE == true ]; then
+        set -x
+    fi
```
Comment (fillup) [inline ecs-deploy]: i agree, I'll move it

File: ecs-deploy (hunk 2 above)
Comment (fillup) [inline ecs-deploy]: sounds like a good idea

File: ecs-deploy (hunk 1 above)
Comment (fillup) [inline ecs-deploy]: sure, I dont see any problems with that

File: Dockerfile [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
 
-RUN apt-get update -y \
-    && apt-get install -y \
-        curl \
-        python-setuptools \
-        jq \
-    && easy_install pip \
-    && pip install awscli
+# Update APK cache
+RUN apk update
```
Comment (jangrewe) [inline Dockerfile]: This is not needed if using `--no-cache` (see below)

File: Dockerfile [hunk 6]
```diff
@@ -1,16 +1,17 @@
...
 
-COPY ecs-deploy /usr/local/bin/ecs-deploy
+# Install packges needed
+RUN apk add ca-certificates curl bash jq py2-pip && \
```
Comment (jangrewe) [inline Dockerfile]: Use `RUN apk add --no-cache ca-certificates curl bash jq py2-pip` to leave no package lists behind.

**Comment by jangrewe** [comment]
It just ran a quick deployment with ecs-deploy in our Docker image replaced by this version, seems to work fine without any further adjustments.

File: ecs-deploy [hunk 7]
```diff
@@ -1,408 +1,466 @@
...
-    while [ $i -lt $TIMEOUT ]
-    do
-      # Scan the list of running tasks for that service, and see if one of them is the
-      # new version of the task definition
+function createNewTaskDefJson() {
+    # Get a JSON representation of the current task definition
+    # + Update definition to use new image name
+    # + Filter the def
+    DEF=$( echo "$TASK_DEFINITION" \
+            | sed -e "s|\"image\": *\"${imageWithoutTag}:.*\"|\"image\": \"${useImage}\"|g" \
```
Comment (gregwebs) [inline ecs-deploy]: I generally prefer to use `jq .+{image:...`. Its definitely still a pain to drill down (although there is probably a jq function for this). Example: https://github.com/KariusDx/aws-scripts/blob/master/ecs-one-off#L91

**Comment by gregwebs** [comment]
You are making shell scripting look like real software engineering :)

**Comment by fillup** [comment]
@jangrewe thanks for testing this out and confirming it works for you.

File: Dockerfile (hunk 6 above)
Comment (fillup) [inline Dockerfile]: ok, I made the updates, saved a meg :-)

File: ecs-deploy (hunk 7 above)
Comment (fillup) [inline ecs-deploy]: We can make changes to make it more robust or stable, this round though I was just restructuring it for functions and testing. But once it is merged in if you have other improvements I'm happy to incorporate them.

**Comment by jangrewe** [comment]
Unfortunately i discovered an issue with this version of the script:

```
$ ecs-deploy -t 300 --max-definitions 5 --cluster ${ECS_CLUSTER} --service-name ${SERVICE}${WORKER} --image ${ECR_REPO}${WORKER}:latest
Using image name: 803064921768.dkr.ecr.eu-west-1.amazonaws.com/registration:latest
Current task definition: arn:aws:ecs:eu-west-1:803064921768:task-definition/registration_blue:132
New task definition: arn:aws:ecs:eu-west-1:803064921768:task-definition/registration_blue:133
Service updated successfully, new task definition running.
usage: aws [options] <command> <subcommand> [<subcommand> ...] [parameters]
To see help text, you can run:

  aws help
  aws <command> help
  aws <command> <subcommand> help

Unknown options: [, {, "memoryReservation":, 300,, "environment":, [, {, "name":, "FLUENT_HOST",, "value":, "172.17.0.1", },, {, "name":, "AWS_ENV",, "value":, "<redacted>", },, {, "name":, "DATABASE_URL",, "value":, "<redacted>", },, {, "name":, "<redacted>",, "value":, "<redacted>", },, {, "name":, "CLUSTER",, "value":, "blue", },, {, "name":, "REDIS_URL",, "value":, "<redacted>, "containerDefinitions":
ERROR: Build failed: exit code 1
```
This task definition is nearly identical to all others we deploy, but has a couple more environment variables. It works fine with the current version from `master`

**Comment by fillup** [comment]
Thanks for letting me know @jangrewe, got busy this week so I'll have to check that out further next week. Looks like that is when it is trying to clear out old task definitions.

**Comment by fillup** [comment]
@jangrewe I figured out what I broke that caused that error for you and got it fixed, so I'll merge in to develop now.

**Comment by jangrewe** [comment]
Thanks, i'll take it for another spin later today!

**Pull request merged**